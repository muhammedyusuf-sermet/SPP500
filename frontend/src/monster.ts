export enum Size {
	Tiny = "Tiny",
	Small = "Small",
	Medium = "Medium",
	Large = "Large",
	Huge = "Huge",
	Gargantuan = "Gargantuan"
}

export enum MonsterType {
	Beast = "Beast",
	Dragon = "Dragon",
	Monstrosity = "Monstrosity",
	Humanoid = "Humanoid",
	Fiend = "Fiend",
	Undead = "Undead",
	Giant = "Giant",
	Elemental = "Elemental",
	SwarmOfTinyBeasts = "SwarmOfTinyBeasts",
	Construct = "Construct",
	Celestial = "Celestial",
	Aberration = "Aberration",
	Fey = "Fey",
	Plant = "Plant",
	Ooze = "Ooze"
}

export enum MonsterRace {
	AnyRace = "AnyRace",
	Devil = "Devil",
	Demon = "Demon",
	Human = "Human",
	Shapechanger = "Shapechanger",
	Goblinoid = "Goblinoid",
	Titan = "Titan",
	Gnoll = "Gnoll",
	Gnome = "Gnome",
	Dwarf = "Dwarf",
	Elf = "Elf",
	Orc = "Orc",
	Kobold = "Kobold",
	Lizardfolk = "Lizardfolk",
	Merfolk = "Merfolk",
	Sahuagin = "Sahuagin",
	Grimlock = "Grimlock"
}

export enum Alignment {
	Unaligned = "Unaligned",
	AnyAlignment = "AnyAlignment",
	AnyNonGoodAlignment = "AnyNonGoodAlignment",
	AnyNonEvilAlignment = "AnyNonEvilAlignment",
	AnyNonLawfulAlignment = "AnyNonLawfulAlignment",
	AnyNonChaoticAlignment = "AnyNonChaoticAlignment",
	AnyGoodAlignment = "AnyGoodAlignment",
	AnyEvilAlignment = "AnyEvilAlignment",
	AnyLawfulAlignment = "AnyLawfulAlignment",
	AnyChaoticAlignment = "AnyChaoticAlignment",
	AnyNeutralAlignment = "AnyNeutralAlignment",
	LawfulGood = "LawfulGood",
	LawfulNeutral = "LawfulNeutral",
	LawfulEvil = "LawfulEvil",
	NeutralGood = "NeutralGood",
	Neutral = "Neutral",
	NeutralEvil = "NeutralEvil",
	ChaoticGood = "ChaoticGood",
	ChaoticNeutral = "ChaoticNeutral",
	ChaoticEvil = "ChaoticEvil"
}

export enum Environment {
	Arctic = "Arctic",
	Coastal = "Coastal",
	Desert = "Desert",
	Forest = "Forest",
	Grassland = "Grassland",
	Hill = "Hill",
	Mountain = "Mountain",
	Swamp = "Swamp",
	Underdark = "Underdark",
	Underwater = "Underwater",
	Urban = "Urban"
}

export enum MonsterAction {
	SpecialAbility = "SpecialAbility",
	Action = "Action",
	LegendaryAction = "LegendaryAction"
}

export interface IMonsterAbilities {
	name: string,
	description: string,
}

export enum TypeOfAction {
	Ranged = "Ranged",
	Melee = "Melee",
}

export interface IMonsterActions {
	name: string,
	description: string,
	usageCount: number,
	legendaryAction: boolean,
	typeOfAction?: TypeOfAction,
	toHit: number,
	reach: number,
	targets: number,
	averagedDamage: number,
	damageDice: number,
	damageConstant: number,

}

export interface IMonsterState {
	Id?: number,
	Name: string,
	Size?: Size | string;
	Type?: MonsterType | string;
	Race?: MonsterRace | string;
	Environment?: Environment | string;
	Alignment?: Alignment | string;

	ArmorClass?: number;
	HitPoints?: number;
	HitPointDistribution?: string;

	Speed?: string;
	Languages?: string;

	DamageVulnerabilities?: string;
	DamageResistances?: string;
	DamageImmunities?: string;
	ConditionImmunities?: string;

	ChallengeRating?: number;

	AbilityScores: IMonsterAbilityScoreState;
	Senses: SenseMap;
	Skills: SkillMap;
	SavingThrows: IMonsterSavingThrowState;
	//Actions: IActionState[];
}

export interface SkillMap {
	[skillName: string]: number|undefined;
}

export interface SenseMap {
	[senseName: string]: number|undefined;
}

export interface IMonsterAbilityScoreState {
	[index:string]: number|undefined;
	Strength?: number;
	Dexterity?: number;
	Constitution?: number;
	Intelligence?: number;
	Wisdom?: number;
	Charisma?: number;
}

export interface IMonsterSavingThrowState {
	[index:string]: number|undefined;
	Strength?: number;
	Dexterity?: number;
	Constitution?: number;
	Intelligence?: number;
	Wisdom?: number;
	Charisma?: number;
}

export interface IActionState {
	Name: string,
	Description: string,
	HitBonus?: number,
	Damage?: string,
	DamageBonus?: number,
	Type?: MonsterAction
}
/*export interface IMonster {
	name: string,
	type: MonsterType,
	alignment: MonsterAlignment,
	size: MonsterSize,
	race: MonsterRace,
	environment: MonsterEnvironment,
	resistance?: string,
	damageImmunity?: string,
	conditionImmunity?: string,
	vulnerability?: string,
	armorClass: number,
	hitPoints?: number,
	hitPointDice?: string,
	hitPointDiceAdd?: number,
	speedLand: number,
	speedSwim?: number,
	strStat: number,
	dexStat: number,
	conStat: number,
	intStat: number,
	wisStat: number,
	chaStat: number,
	strSavingThrow?: number,
	dexSavingThrow?: number,
	conSavingThrow?: number,
	intSavingThrow?: number,
	wisSavingThrow?: number,
	chaSavingThrow?: number,
	skillsAthletics: number,
	skillsAcrobatics?: number,
	skillsSleightOfHand?: number,
	skillsStealth?: number,
	skillsArcana?: number,
	skillsHistory?: number,
	skillsInvestigation?: number,
	skillsNature?: number,
	skillsReligion?: number,
	skillsAnimalHandling?: number,
	skillsInsight?: number,
	skillsMedicine?: number,
	skillsPerception?: number,
	skillsSurvival?: number,
	skillsDeception?: number,
	skillsIntimidation?: number,
	skillsPerformance?: number,
	skillsPersuasion?: number,
	sensesBlindsight?: number,
	sensesDarkvision?: number,
	sensesTremorsense?: number,
	sensesTruesight?: number,
	sensesPassivePerception?: number,
	sensesPassiveInvestigation?: number,
	sensesPassiveInsight?: number,
	languages?: string,
	challengeRating: number,
	experiencePoints: number,
	abilities?: IMonsterAbilities[],
	actions?: IMonsterActions[],
}*/