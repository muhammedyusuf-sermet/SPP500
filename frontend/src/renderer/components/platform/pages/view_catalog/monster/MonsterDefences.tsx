import * as React from 'react';
const Joi = require('joi');
import { ValidationError, ValidationOptions, JoiObject } from 'joi';

import 'bulma/css/bulma.css';

import { isDeepStrictEqual } from 'util';
import { ExpansionPanel, ExpansionPanelSummary, ExpansionPanelDetails, Typography, FormControl, InputLabel, Input, FormHelperText, Grid } from '@material-ui/core';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';

export interface IMonsterDefencesProps {
	disabled?: boolean,
	// validation
	PayloadSchema: JoiObject,
	ValidationOptions: ValidationOptions,
	// initial values
	initial: {
		ArmorClass?: number;
		HitPoints?: number;
		HitPointDistribution?: string;
	}
}

export interface IMonsterDefencesState {
	// data
	ArmorClass?: number;
	HitPoints?: number;
	HitPointDistribution?: string;
	// errors
	ArmorClassError?: string;
	HitPointsError?: string;
	HitPointDistributionError?: string;
}

export class MonsterDefences extends React.Component<IMonsterDefencesProps, IMonsterDefencesState> {
	constructor(props: IMonsterDefencesProps) {
		super(props);
		this.state = {
			...props.initial
		};
	}

	componentWillReceiveProps(nextProps: IMonsterDefencesProps) {
		if (isDeepStrictEqual(this.props.initial, nextProps.initial) == false) {
			this.setState({
				...nextProps.initial
			});
		} else if (this.props.disabled != nextProps.disabled) {
			this.forceUpdate();
		}
	}

	stringToNumber = (toConvert : string) => {
		return isNaN(parseInt(toConvert)) ? undefined : parseInt(toConvert);
	}

	handleNumberChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const value = this.stringToNumber(event.target.value);
		const name = event.currentTarget.name;
		Joi.validate(
			value,
			Joi.reach(this.props.PayloadSchema, [name]),
			this.props.ValidationOptions,
			(errors: ValidationError) => {
				this.setState({
					[name]: value,
					[name+'Error']: errors ? errors.details[0].message : undefined
				});
			});
	}

	handleStringChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		let value: string|undefined = event.target.value;
		if (/^\s*$/.test(value))
			value = undefined
		Joi.validate(
			value,
			Joi.reach(this.props.PayloadSchema, [event.currentTarget.name]),
			this.props.ValidationOptions,
			(errors: ValidationError) => {
				this.setState({
					[event.currentTarget.name]: value,
					[event.currentTarget.name+'Error']: errors ? errors.details[0].message : undefined
				});
			});
	}

	render() {
		return (
			<ExpansionPanel defaultExpanded CollapseProps={{timeout: 100}} >
				<ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
					<Typography className='heading' >Armor Class and Hit Points</Typography>
				</ExpansionPanelSummary>
				<ExpansionPanelDetails>
					<Grid container spacing={8} >
						<Grid item xs={12}>
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor="ArmorClass">Armor Class</InputLabel>
								<Input
									id='ArmorClass'
									type='number'
									value={this.state.ArmorClass != undefined ? this.state.ArmorClass : ''}
									name='ArmorClass'
									onChange={this.handleNumberChange}
									aria-describedby="ArmorClass-helper-text" />
								<FormHelperText error id="ArmorClass-helper-text">{this.state.ArmorClassError}</FormHelperText>
							</FormControl>
						</Grid>
						<Grid item xs={6}>
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor="HitPoints">Hit Points</InputLabel>
								<Input
									id='HitPoints'
									type='number'
									value={this.state.HitPoints != undefined ? this.state.HitPoints : ''}
									name='HitPoints'
									onChange={this.handleNumberChange}
									aria-describedby="HitPoints-helper-text" />
								<FormHelperText error id="HitPoints-helper-text">{this.state.HitPointsError}</FormHelperText>
							</FormControl>
						</Grid>
						<Grid item xs={6}>
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor="HitPointDistribution">Hit Point Distribution</InputLabel>
								<Input
									id='HitPointDistribution'
									type='text'
									value={this.state.HitPointDistribution || ''}
									name='HitPointDistribution'
									onChange={this.handleStringChange}
									aria-describedby="HitPointDistribution-helper-text" />
								<FormHelperText error id="HitPointDistribution-helper-text">{this.state.HitPointDistributionError}</FormHelperText>
							</FormControl>
						</Grid>
					</Grid>
				</ExpansionPanelDetails>
			</ExpansionPanel>
		);
	}
}