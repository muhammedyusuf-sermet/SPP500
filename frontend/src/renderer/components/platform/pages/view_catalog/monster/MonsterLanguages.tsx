import * as React from 'react';
const Joi = require('joi');
import { ValidationError, ValidationOptions, JoiObject } from 'joi';

import 'bulma/css/bulma.css';
import { Field, FieldLabel, Label, FieldBody, Control, Input, Help } from 'bloomer';
import { isDeepStrictEqual } from 'util';

export interface IMonsterLanguagesProps {
	disabled?: boolean,
	// validation
	PayloadSchema: JoiObject,
	ValidationOptions: ValidationOptions,
	// initial values
	initial: {
		Languages?: string;
	}
}

export interface IMonsterLanguagesState {
	// data
	Languages?: string;
	// errors
	LanguagesError?: string;
}

export class MonsterLanguages extends React.Component<IMonsterLanguagesProps, IMonsterLanguagesState> {
	constructor(props: IMonsterLanguagesProps) {
		super(props);
		this.state = {
			...props.initial
		};
	}

	componentWillReceiveProps(nextProps: IMonsterLanguagesProps) {
		if (isDeepStrictEqual(this.props.initial, nextProps.initial) == false) {
			this.setState({
				...nextProps.initial
			});
		} else if (this.props.disabled != nextProps.disabled) {
			this.forceUpdate();
		}
	}

	handleLanguagesChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		let value: string|undefined = event.target.value;
		if (/^\s*$/.test(value))
			value = undefined
		Joi.validate(
			value,
			Joi.reach(this.props.PayloadSchema, [event.currentTarget.name]),
			this.props.ValidationOptions,
			(errors: ValidationError) => {
				this.setState({
					[event.currentTarget.name]: value,
					[event.currentTarget.name+'Error']: errors ? errors.details[0].message : undefined
				});
			});
	}

	render() {
		return (
			<Field isHorizontal>
				<FieldLabel isNormal>
					<Label>Languages</Label>
				</FieldLabel>
				<FieldBody>
					<Field>
						<Control>
							<Input
								disabled={this.props.disabled}
								id='Languages'
								type='text'
								placeholder='Languages'
								autoComplete='Languages'
								value={this.state.Languages || ''}
								name='Languages'
								onChange={this.handleLanguagesChange} />
						</Control>
						<Help id='Languages' isColor='danger'>{this.state.LanguagesError}</Help>
					</Field>
				</FieldBody>
			</Field>
		);
	}
}