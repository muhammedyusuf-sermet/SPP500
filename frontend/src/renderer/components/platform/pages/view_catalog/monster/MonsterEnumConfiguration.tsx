import * as React from 'react';
const Joi = require('joi');
import { ValidationError, ValidationOptions, JoiObject } from 'joi';

import 'bulma/css/bulma.css';

import { MonsterType, MonsterRace, Size, Environment, Alignment } from '../../../../../../monster';
import { isDeepStrictEqual } from 'util';
import { ExpansionPanel, ExpansionPanelSummary, ExpansionPanelDetails, Typography, FormControl, InputLabel, Input, FormHelperText, Grid, Select, Tooltip } from '@material-ui/core';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import {Help} from "bloomer";

const types = Object.values(MonsterType);
const sizes = Object.values(Size);
const races = Object.values(MonsterRace);
const environments = Object.values(Environment);
const alignments = Object.values(Alignment);

export interface IMonsterDropdownProps {
	disabled?: boolean,
	name: string,
	selected?: string,
	options: string[]
	onChange: (name: string, selectOption?: string) => void
}

export interface IMonsterDropdownState {
	selected: string
}

export class MonsterDropdown extends React.Component<IMonsterDropdownProps, IMonsterDropdownState> {
	constructor(props: IMonsterDropdownProps) {
		super(props);
		this.state = {
			selected: props.selected ? props.selected : 'Default'
		};
	}

	componentWillReceiveProps(nextProps: IMonsterDropdownProps) {
		if (this.state.selected != nextProps.selected)
			this.setState({
				selected: nextProps.selected ? nextProps.selected : 'Default'
			});
	}

	handleChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
		this.setState({selected: event.target.value});
		this.props.onChange(this.props.name, event.target.value);
	}

	render() {
		return (
			<Select
				native={true}
				id={this.props.name}
				disabled={this.props.disabled}
				onChange={this.handleChange}
				value={this.state.selected}
				input={<Input name={this.props.name} id={this.props.name} />} >
				<option
					id={this.props.name+'.'+'Default'}
					value='' >
					Default
				</option>
				{this.props.options.map(option =>
					<option
						key={this.props.name+'.'+option}
						id={this.props.name+'.'+option}
						value={option}>
						{option.replace(/([A-Z])/g, ' $1').trim()}
					</option>
				)}
			</Select>
		);
	}
}

export interface IMonsterEnumConfigurationProps {
	disabled?: boolean,
	// validation
	PayloadSchema: JoiObject,
	ValidationOptions: ValidationOptions,
	// initial values
	initial: {
		Size?: string;
		Type?: string;
		Race?: string;
		Alignment?: string;
		Environment?: string;
	}
}

export interface IMonsterEnumConfigurationState {
	// data
	Size?: string;
	Type?: string;
	Race?: string;
	Alignment?: string;
	Environment?: string;
	// errors
	SizeError?: string;
	TypeError?: string;
	RaceError?: string;
	AlignmentError?: string;
	EnvironmentError?: string;
}

export class MonsterEnumConfiguration extends React.Component<IMonsterEnumConfigurationProps, IMonsterEnumConfigurationState> {
	constructor(props: IMonsterEnumConfigurationProps) {
		super(props);
		this.state = {
			...props.initial
		};
	}

	componentWillReceiveProps(nextProps: IMonsterEnumConfigurationProps) {
		if (isDeepStrictEqual(this.props.initial, nextProps.initial) == false)
			this.setState({
				...nextProps.initial
			});
	}

	handleMonsterEnumChange = (name: string, newEnum?: string) => {
		if (newEnum) {
			Joi.validate(
				newEnum,
				Joi.reach(this.props.PayloadSchema, [name]),
				this.props.ValidationOptions,
				(errors: ValidationError) => {
					if(errors) {
						this.setState({
							[name+'Error']: errors.details[0].message
						});
					} else {
						this.setState({
							[name]: newEnum,
							[name+'Error']: undefined
						});
					}
				});
		} else {
			this.setState({
				[name]: undefined,
				[name+'Error']: undefined
			});
		}
	}

	render() {
		return (
			<ExpansionPanel defaultExpanded CollapseProps={{timeout: 100}} >
				<ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
					<Typography className='heading' >Basic Configurations</Typography>
				</ExpansionPanelSummary>
				<ExpansionPanelDetails>
					<Grid container spacing={8} >
						<Grid item xs={4} >
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor='Type' >Type</InputLabel>
								<MonsterDropdown
									disabled={this.props.disabled}
									selected={this.state.Type}
									name='Type'
									options={types}
									onChange={this.handleMonsterEnumChange} />
								<FormHelperText error id='Type-helper-text' >{this.state.TypeError}</FormHelperText>
								<Tooltip disableFocusListener title={"Creature types are rough categories of creatures which determine the way game mechanics affect the creature. The choice of type is important, as all creatures which have a given type will share certain characteristics (with some exceptions)."}>
										<Help>Hover for More About Monster Type</Help>
								</Tooltip>
							</FormControl>
						</Grid>
						<Grid item xs={4} >
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor='Size' >Size</InputLabel>
								<MonsterDropdown
									disabled={this.props.disabled}
									selected={this.state.Size}
									name='Size'
									options={sizes}
									onChange={this.handleMonsterEnumChange} />
								<FormHelperText error id='Size-helper-text' >{this.state.SizeError}</FormHelperText>
								<Tooltip disableFocusListener title={"A creature's size determines how much space it occupies on squares or hexes, as shown in the Creature Size and Space table. If the miniature you use for a monster takes up an amount of space different from what's on the table, that's fine, but treat the monster as its official size for all other rules."}>
										<Help>Hover for More About Monster Size</Help>
								</Tooltip>
							</FormControl>
						</Grid>
						<Grid item xs={4} >
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor='Race' >Race</InputLabel>
								<MonsterDropdown
									disabled={this.props.disabled}
									selected={this.state.Race}
									name='Race'
									options={races}
									onChange={this.handleMonsterEnumChange} />
								<FormHelperText error id='Race-helper-text' >{this.state.RaceError}</FormHelperText>
								<Tooltip disableFocusListener title={"A character race is a fundamental part of the identity and nature of characters in the Dungeons & Dragons role-playing game. Each race has a distinct appearance, behavior and often range of statistics associated with it. See your DM handbook for more information."}>
										<Help>Hover for More About Monster Race</Help>
								</Tooltip>
							</FormControl>
						</Grid>
						<Grid item xs={6} >
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor='Alignment' >Alignment</InputLabel>
								<MonsterDropdown
									disabled={this.props.disabled}
									selected={this.state.Alignment}
									name='Alignment'
									options={alignments}
									onChange={this.handleMonsterEnumChange} />
								<FormHelperText error id='Alignment-helper-text' >{this.state.AlignmentError}</FormHelperText>
								<Tooltip disableFocusListener title={"Alignment is a categorization of the ethical and moral perspective of player characters, non-player characters, and creatures. Most versions of the game feature a system in which players make two choices for characters. One is the character's views on 'law' vs 'chaos', the other on 'good' vs 'evil'."}>
										<Help>Hover for More About Monster Alignment</Help>
								</Tooltip>
							</FormControl>
						</Grid>
						<Grid item xs={6} >
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor='Environment' >Environment</InputLabel>
								<MonsterDropdown
									disabled={this.props.disabled}
									selected={this.state.Environment}
									name='Environment'
									options={environments}
									onChange={this.handleMonsterEnumChange} />
								<FormHelperText error id='Environment-helper-text' >{this.state.EnvironmentError}</FormHelperText>
								<Tooltip disableFocusListener title={"Monster environment is the type of environment that the monster is most likely to be found in."}>
										<Help>Hover for More About Monster Environment</Help>
								</Tooltip>
							</FormControl>
						</Grid>
					</Grid>
				</ExpansionPanelDetails>
			</ExpansionPanel>
		);
	}
}