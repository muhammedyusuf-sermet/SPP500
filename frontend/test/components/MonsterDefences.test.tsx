import * as React from 'react';
import { mount, ReactWrapper } from 'enzyme';

import { IMonsterDefencesProps, IMonsterDefencesState, MonsterDefences } from '../../src/renderer/components/platform/pages/view_catalog/monster/MonsterDefences';
const Joi = require('joi');
import { ValidationOptions } from 'joi';

const payloadSchema = Joi.object({
	ArmorClass: Joi.number().integer().greater(0).label('ArmorClass'),
	HitPoints: Joi.number().integer().greater(0).label('HitPoints'),
	// (rolls 'd' dice [+ - * /] operation) one or more times then rolls 'd' dice
	HitPointDistribution: Joi.string().max(20).regex(/^((\d+d\d+)[\+\-\*\/])*(\d+d\d+)([\+\-\*\/]\d+)?$/, '#d# OR (#d# operator (#d# or number)) NO spaces').label('HitPointDistribution'),
});
const validateOptions: ValidationOptions = {
	abortEarly: false,
	convert: true,
	allowUnknown: false,
	context: {

	}
};

////// Happy Path //////
describe('Monster Defences', () => {

	let monsterDefencesInstance: ReactWrapper<IMonsterDefencesProps, IMonsterDefencesState, MonsterDefences>;

	describe('Provided Resistance values', () => {

		beforeEach(() => {
			monsterDefencesInstance =
				mount<MonsterDefences, IMonsterDefencesProps, IMonsterDefencesState>(
					<MonsterDefences
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						initial={{
							ArmorClass: 124,
							HitPoints: 623,
							HitPointDistribution: '5d12+34'
						}} />
				);
		})

		it('renders without crashing', () => {
			expect(monsterDefencesInstance).toBeDefined();
		});

		it('should have the correct number for ArmorClass', () => {
			expect(monsterDefencesInstance.find('Input#ArmorClass').props().value).toEqual(124);
		});

		it('should have the correct number for HitPoints', () => {
			expect(monsterDefencesInstance.find('Input#HitPoints').props().value).toEqual(623);
		});

		it('should have the correct string for HitPointDistribution', () => {
			expect(monsterDefencesInstance.find('Input#HitPointDistribution').props().value).toEqual('5d12+34');
		});
	});

	describe('Happy Path', () => {

		beforeEach(() => {
			monsterDefencesInstance =
				mount<MonsterDefences, IMonsterDefencesProps, IMonsterDefencesState>(
					<MonsterDefences
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						initial={{}}/>
				);
		})

		it('renders without crashing', () => {
			expect(monsterDefencesInstance).toBeDefined();
		});

		it('should be able to update state', () => {
			monsterDefencesInstance.find('Input#ArmorClass').simulate('change', { target: { value: '43' } })
			monsterDefencesInstance.find('Input#HitPoints').simulate('change', { target: { value: '564' } })
			monsterDefencesInstance.find('Input#HitPointDistribution').simulate('change', { target: { value: '3d6+92' } })
			expect(monsterDefencesInstance.state()).toEqual({
				// data
				ArmorClass: 43,
				HitPoints: 564,
				HitPointDistribution: '3d6+92',
				// errors
				ArmorClassError: undefined,
				HitPointsError: undefined,
				HitPointDistributionError: undefined,
			});
		});
		// STATE CHANGE
		it('should change only ArmorClass when ArmorClass is changed', () => {
			monsterDefencesInstance.find('Input#ArmorClass').simulate('change', { target: { value: '958' } })
			expect(monsterDefencesInstance.state().ArmorClass).toEqual(958);
		})

		it('should change only HitPoints when HitPoints is changed', () => {
			monsterDefencesInstance.find('Input#HitPoints').simulate('change', { target: { value: '372' } })
			expect(monsterDefencesInstance.state().HitPoints).toEqual(372);
		})

		it('should change only HitPointDistribution when HitPointDistribution is changed', () => {
			monsterDefencesInstance.find('Input#HitPointDistribution').simulate('change', { target: { value: '2d8+73' } })
			expect(monsterDefencesInstance.state().HitPointDistribution).toEqual('2d8+73');
		})
		// DEFAULTS
		it('should change Type back to undefined when input is empty', () => {
			monsterDefencesInstance.find('Input#ArmorClass').simulate('change', { target: { value: '' } })
			expect(monsterDefencesInstance.state().ArmorClass).toEqual(undefined);
		})

		it('should change Alignment back to undefined when input is empty', () => {
			monsterDefencesInstance.find('Input#HitPoints').simulate('change', { target: { value: '' } })
			expect(monsterDefencesInstance.state().HitPoints).toEqual(undefined);
		})

		it('should change Size back to undefined when input is empty', () => {
			monsterDefencesInstance.find('Input#HitPointDistribution').simulate('change', { target: { value: '' } })
			expect(monsterDefencesInstance.state().HitPointDistribution).toEqual(undefined);
		})
		// VALIDATION
		const moreThan20: string = "1000d1000+1000d1000+1000d1000"

		it('should validate ArmorClass when ArmorClass is changed', () => {
			monsterDefencesInstance.find('Input#ArmorClass').simulate('change', { target: { value: '-12'  } })
			expect(monsterDefencesInstance.state().ArmorClass).toEqual(-12);
			expect(monsterDefencesInstance.find('Help#ArmorClass').text()).toEqual('\"ArmorClass\" must be greater than 0');
		})

		it('should validate HitPoints when HitPoints is changed', () => {
			monsterDefencesInstance.find('Input#HitPoints').simulate('change', { target: { value: '-23' } })
			expect(monsterDefencesInstance.state().HitPoints).toEqual(-23);
			expect(monsterDefencesInstance.find('Help#HitPoints').text()).toEqual('\"HitPoints\" must be greater than 0');
		})

		it('should validate HitPointDistribution when HitPointDistribution is changed length', () => {
			monsterDefencesInstance.find('Input#HitPointDistribution').simulate('change', { target: { value: moreThan20 } })
			expect(monsterDefencesInstance.state().HitPointDistribution).toEqual(moreThan20);
			expect(monsterDefencesInstance.find('Help#HitPointDistribution').text()).toEqual('\"HitPointDistribution\" length must be less than or equal to 20 characters long');
		})

		it('should validate HitPointDistribution when HitPointDistribution is changed pattern', () => {
			monsterDefencesInstance.find('Input#HitPointDistribution').simulate('change', { target: { value: 'InvalidPattern' } })
			expect(monsterDefencesInstance.state().HitPointDistribution).toEqual('InvalidPattern');
			expect(monsterDefencesInstance.find('Help#HitPointDistribution').text()).toEqual('\"HitPointDistribution\" with value \"InvalidPattern\" fails to match the #d# OR (#d# operator (#d# or number)) NO spaces pattern');
		})
	});
});