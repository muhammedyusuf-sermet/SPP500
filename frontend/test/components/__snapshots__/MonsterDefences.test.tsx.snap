// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Monster Defences Small snapshot matches snapshot 1`] = `
<WithStyles(ExpansionPanel)
  CollapseProps={
    Object {
      "timeout": 100,
    }
  }
  defaultExpanded={true}
>
  <WithStyles(ExpansionPanelSummary)
    expandIcon={<pure(ExpandMoreIcon) />}
  >
    <WithStyles(Typography)
      className="heading"
    >
      Armor Class and Hit Points
    </WithStyles(Typography)>
  </WithStyles(ExpansionPanelSummary)>
  <WithStyles(ExpansionPanelDetails)>
    <WithStyles(Grid)
      container={true}
      spacing={8}
    >
      <WithStyles(Grid)
        item={true}
        xs={12}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="ArmorClass"
          >
            Armor Class
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="ArmorClass-helper-text"
            id="ArmorClass"
            name="ArmorClass"
            onChange={[Function]}
            type="number"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="ArmorClass-helper-text"
          />
          <WithStyles(Tooltip)
            disableFocusListener={true}
            title="Your Armor Class (AC) represents how hard it is for opponents to land a solid, damaging blow on you. Itâ€™s the attack roll result that an opponent needs to achieve to hit you. Your AC is equal to the following: 10 + armor bonus + shield bonus + Dexterity modifier + size modifier"
          >
            <withHelpersModifiers(Help)
              className="tooltip"
            >
              Hover for More About Armor Class
            </withHelpersModifiers(Help)>
          </WithStyles(Tooltip)>
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={6}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="HitPoints"
          >
            Hit Points
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="HitPoints-helper-text"
            id="HitPoints"
            name="HitPoints"
            onChange={[Function]}
            type="number"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="HitPoints-helper-text"
          />
          <WithStyles(Tooltip)
            disableFocusListener={true}
            title="Hit points represent a combination of physical and mental durability, the will to live, and luck. Creatures with more hit points are more difficult to kill. Those with fewer hit points are more fragile. A creature's current hit points (usually just called hit points) can be any number from the creature's hit point maximum down to 0. This number changes frequently as a creature takes damage or receives healing."
          >
            <withHelpersModifiers(Help)
              className="tooltip"
            >
              Hover for More About Hit Points
            </withHelpersModifiers(Help)>
          </WithStyles(Tooltip)>
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={6}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="HitPointDistribution"
          >
            Hit Point Distribution
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="HitPointDistribution-helper-text"
            id="HitPointDistribution"
            name="HitPointDistribution"
            onChange={[Function]}
            type="text"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="HitPointDistribution-helper-text"
          />
          <WithStyles(Tooltip)
            disableFocusListener={true}
            title="For randomized and dynamic encounters, you may choose to define the hit points of a monster by a hit point distribution. This value is how you calculate the monsters hit points at the start of the encounter. It is provided in # of dice * # of sides of dice + modifier in the normal #d#+# format."
          >
            <withHelpersModifiers(Help)
              className="tooltip"
            >
              Hover for More About Hit Point Distribution
            </withHelpersModifiers(Help)>
          </WithStyles(Tooltip)>
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </WithStyles(ExpansionPanelDetails)>
</WithStyles(ExpansionPanel)>
`;
